bin/civihr-install

```diff
@@ -0,0 +1,823 @@
+      \$result .= \$alphabet{rand(1, \$alphabetSize) - 1};
+    }
+    return \$result;
+  }
+  if (empty(\$argv[1])) {
+    \$len = 8;
+  } else {
+    \$len = \$argv[1];
+  }
+  print createRandom(\$len, 'abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ0123456789');
+?>
+EOF
+# Calling temp make password php file.
+php "$TMPDIR/mkpasswd.php" "$1"
+}
+
+###############################################################################
+## Assert that shell variables are defined. If not defined, exit with an error.
+##  If a variable missing, the application dies.
+function civihr_assertvars() {
+  _civihr_assertvars_back="$-"
+
+  set +x
+  context="$1"
+  shift
+  while [ "$1" ]; do
+    var="$1"
+    eval "val=\$$var"
+    if [ -z "$val" ]; then
+      echo "missing variable: $var [in $context]"
+      exit 98
+    fi
+    shift
+  done
+
+  set -${_civihr_assertvars_back} 
+}
+
+###############################################################################
+## Summarize the content of key environment variables
+function civihr_summary() {
+  if [ -n "$1" ]; then
+    echo "$1"
+  fi
+  shift
+
+  civihr_assertvars civihr_summary "$@"
+  for var in "$@" ; do
+    eval "val=\$$var"
+    echo " - $var: $val"
+  done
+}
+
+
+###############################################################################
+## Creates and ensure parent directory exists.
+## civihr_makeparent <file>
+function civihr_makeparent() {
+  parentdir=$(dirname "$1")
+  if [ ! -d "$parentdir" ]; then
+    mkdir -p "$parentdir"
+  fi
+}
+
+###############################################################################
+## Checks whether directory exists, if not then create it.
+function civihr_mkdir() {
+  for a in "$@" ; do
+    if [ ! -d "$a" ]; then
+      mkdir -p "$a"
+    fi
+  done
+}
+
+###############################################################################
+## Creates host and port to a single string.
+function civihr_build_hostport() {
+  local host=$1
+  local port=$2
+  if [ -z "$port" ]; then
+    echo "$host"
+  else
+    echo "$host:$port"
+  fi
+}
+
+###############################################################################
+## Making Random temp files
+function civihr_maketempfile(){
+ cat > "$TMPDIR/mktemp.php" << EOF
+<?php
+    \$prefix = empty(\$argv[1]) ? 'mktemp-' : \$argv[1];
+\$basedir = empty(\$argv[2]) ? sys_get_temp_dir() : \$argv[2];
+echo tempnam(\$basedir,  \$prefix);
+?>
+EOF
+}
+
+###############################################################################
+## Setup MySQL services
+function database_install() {
+  _database_install_cms
+  _database_install_civi
+}
+
+## Setting up Database for CMS i.e Drupal
+function _database_install_cms() {
+
+  echo "[[Setting up MySQL for CMS]]"
+  civihr_assertvars _database_install_cms CMS_ROOT SITE_NAME SITE_ID TMPDIR
+  # Create database.
+  mysql -u$DB_USER -p$DB_PASS -e "create database $CMSdbName" || { echo "[[Please fix above MySQL error.]]"; exit 1;  }
```

Comment:
I stumbled upon a very annoying error: If an error happens at the middle of the installation, but after the database has been created, in order to run it again, I'll have to go to MySQL and manually drop the database the script created, otherwise I won't be able to proceed because this line will fail since the database already exists.

I can think of 2 possible improvements here:
- Implement some sort of rollback in case of errors (maybe using traps). Not sure of how difficult/complex it would be to implement this
- Add an option to force droping existing databases during install